File: src/main/java/com/coresql/tokenizer/Tokenizer.java
package com.coresql.tokenizer;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public class Tokenizer {
    private static final Set<String> KEYWORDS = Set.of(
        "CREATE", "TABLE", "INSERT", "INTO", "VALUES", "SELECT", "FROM", "WHERE"
    );

    private final String query;
    private int pos = 0;

    public Tokenizer(String query) {
        this.query = query;
    }

    private char peek() {
        if (isEof()) return '\0';
        return query.charAt(pos);
    }

    private char advance() {
        if (isEof()) return '\0';
        return query.charAt(pos++);
    }

    private boolean isEof() {
        return pos >= query.length();
    }

    private void skipWhitespace() {
        while (!isEof() && Character.isWhitespace(peek())) {
            advance();
        }
    }

    private Token readNumber() {
        StringBuilder val = new StringBuilder();
        while (!isEof() && Character.isDigit(peek())) {
            val.append(advance());
        }
        return new Token(TokenType.INTEGER, val.toString());
    }

    private Token readString() {
        advance(); // skip opening quote
        StringBuilder val = new StringBuilder();
        while (!isEof() && peek() != '"' && peek() != '\'') {
            val.append(advance());
        }
        if (!isEof()) advance(); // skip closing quote
        return new Token(TokenType.STRING, val.toString());
    }

    private Token readIdentifierOrKeyword() {
        StringBuilder val = new StringBuilder();
        while (!isEof() && (Character.isLetterOrDigit(peek()) || peek() == '_')) {
            val.append(advance());
        }
        
        String upperVal = val.toString().toUpperCase();
        if (KEYWORDS.contains(upperVal)) {
            return new Token(TokenType.KEYWORD, upperVal);
        }
        return new Token(TokenType.IDENTIFIER, val.toString());
    }

    private Token readOperatorOrSymbol() {
        char c = advance();
        if (c == '=' || c == '<' || c == '>') {
            return new Token(TokenType.OPERATOR, String.valueOf(c));
        } else if (c == '(' || c == ')' || c == ',' || c == ';' || c == '*') {
            return new Token(TokenType.SYMBOL, String.valueOf(c));
        }
        return new Token(TokenType.UNKNOWN, String.valueOf(c));
    }

    public List<Token> tokenize() {
        List<Token> tokens = new ArrayList<>();
        
        while (!isEof()) {
            skipWhitespace();
            if (isEof()) break;

            char c = peek();
            if (Character.isLetter(c) || c == '_') {
                tokens.add(readIdentifierOrKeyword());
            } else if (Character.isDigit(c)) {
                tokens.add(readNumber());
            } else if (c == '"' || c == '\'') {
                tokens.add(readString());
            } else {
                tokens.add(readOperatorOrSymbol());
            }
        }
        tokens.add(new Token(TokenType.END_OF_FILE, ""));
        return tokens;
    }
}

File: src/tokenizer/tokenizer.cpp
#include "tokenizer.h"
#include <cctype>
#include <unordered_set>
#include <algorithm>

const std::unordered_set<std::string> KEYWORDS = {
    "CREATE", "TABLE", "INSERT", "INTO", "VALUES", "SELECT", "FROM", "WHERE"
};

Tokenizer::Tokenizer(const std::string& query) : query(query), pos(0) {}

char Tokenizer::peek() const {
    if (is_eof()) return '\0';
    return query[pos];
}

char Tokenizer::advance() {
    if (is_eof()) return '\0';
    return query[pos++];
}

bool Tokenizer::is_eof() const {
    return pos >= query.length();
}

void Tokenizer::skip_whitespace() {
    while (!is_eof() && std::isspace(peek())) {
        advance();
    }
}

Token Tokenizer::read_number() {
    std::string val = "";
    while (!is_eof() && std::isdigit(peek())) {
        val += advance();
    }
    return {TokenType::INTEGER, val};
}

Token Tokenizer::read_string() {
    advance(); // skip opening quote
    std::string val = "";
    while (!is_eof() && peek() != '"' && peek() != '\'') {
        val += advance();
    }
    if (!is_eof()) advance(); // skip closing quote
    return {TokenType::STRING, val};
}

Token Tokenizer::read_identifier_or_keyword() {
    std::string val = "";
    while (!is_eof() && (std::isalnum(peek()) || peek() == '_')) {
        val += advance();
    }
    
    // Check if keyword (case-insensitive comparison using uppercase)
    std::string upper_val = val;
    std::transform(upper_val.begin(), upper_val.end(), upper_val.begin(), ::toupper);
    
    if (KEYWORDS.find(upper_val) != KEYWORDS.end()) {
        return {TokenType::KEYWORD, upper_val};
    }
    return {TokenType::IDENTIFIER, val};
}

Token Tokenizer::read_operator_or_symbol() {
    char c = advance();
    if (c == '=' || c == '<' || c == '>') {
        return {TokenType::OPERATOR, std::string(1, c)};
    } else if (c == '(' || c == ')' || c == ',' || c == ';') {
        return {TokenType::SYMBOL, std::string(1, c)};
    }
    return {TokenType::UNKNOWN, std::string(1, c)};
}

std::vector<Token> Tokenizer::tokenize() {
    std::vector<Token> tokens;
    
    while (!is_eof()) {
        skip_whitespace();
        if (is_eof()) break;

        char c = peek();
        if (std::isalpha(c) || c == '_') {
            tokens.push_back(read_identifier_or_keyword());
        } else if (std::isdigit(c)) {
            tokens.push_back(read_number());
        } else if (c == '"' || c == '\'') {
            tokens.push_back(read_string());
        } else {
            tokens.push_back(read_operator_or_symbol());
        }
    }
    tokens.push_back({TokenType::END_OF_FILE, ""});
    return tokens;
}
